### PR TITLE
Ranking losses, normalization and pooling layers.

### PR DESCRIPTION
**Pull request by ykate1998**
# Description

This commit introduces supporting modules (ranking losses, normalization and pooling layers) for the implementation of the following papers:

**Fine-tuning CNN Image Retrieval with No Human Annotation**,  
Radenović F., Tolias G., Chum O., 
TPAMI 2018 [[arXiv](https://arxiv.org/abs/1711.02512)]

**CNN Image Retrieval Learns from BoW: Unsupervised Fine-Tuning with Hard Examples**,  
Radenović F., Tolias G., Chum O., 
ECCV 2016 [[arXiv](http://arxiv.org/abs/1604.02426)]

## Type of change

Added ranking losses, normalization and pooling layers.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] A new research paper code implementation

## Tests

Code is compatible with the DELF library. For every implemented function we provide tests in the corresponding modules.

### CODE REVIEW

**Review by andrefaraujo — APPROVED** [review]
Thanks Kate!

File: research/delf/delf/python/training/model/resnet50.py [hunk 1]
```diff
@@ -54,23 +55,23 @@ def __init__(self, kernel_size, filters, stage, block, data_format):
     bn_axis = 1 if data_format == 'channels_first' else 3
 
     self.conv2a = layers.Conv2D(
-        filters1, (1, 1), name=conv_name_base + '2a', data_format=data_format)
+      filters1, (1, 1), name=conv_name_base + '2a', data_format=data_format)
```
Comment (dan-anghel) [inline research/delf/delf/python/training/model/resnet50.py:58]: I see these padding changes in the method arguments from 4 spaces to 2 spaces from line 57 to line 298 are not consistent with our internal codebase. Is this expected?

File: research/delf/delf/python/training/model/resnet50.py [hunk 2]
```diff
@@ -456,27 +457,4 @@ def log_weights(self):
...
-
-  Returns:
-    pooled_feature_map: Tensor of shape [batch, channels].
-  """
-  return tf.pow(
-      tf.reduce_mean(tf.pow(tf.maximum(feature_map, threshold), power),
-                     axis=axis,
-                     keepdims=False),
-      1.0 / power)
+                     layer.name)
```
Comment (dan-anghel) [inline research/delf/delf/python/training/model/resnet50.py:460]: Similarly to the comment above, the missing new line at the end of the file is inconsistent with our internal codebase. Is this expected?

**Review by dan-anghel — COMMENTED** [review]
Thanks André! I just made a few minor comments regarding possible inconsistencies with our internal codebase.
